An error is raised when using pull-through caching in a disconnected environment

This happens when using the pull-through caching workflow with no internet connection.

Traceback:
```
pulp [None]: backoff:ERROR: Giving up download_wrapper(...) after 4 tries (aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known])
[2024-02-03 17:20:47 +0000] [6911] [ERROR] Error handling request
Traceback (most recent call last):
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/connector.py", line 1173, in _create_direct_connection
    hosts = await asyncio.shield(host_resolved)
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/connector.py", line 884, in _resolve_host
    addrs = await self._resolver.resolve(host, port, family=self._family)
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/resolver.py", line 33, in resolve
    infos = await self._loop.getaddrinfo(
  File "/usr/lib64/python3.8/asyncio/base_events.py", line 825, in getaddrinfo
    return await self.run_in_executor(
  File "/usr/lib64/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/lib64/python3.8/socket.py", line 918, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/web_protocol.py", line 452, in _handle_request
    resp = await request_handler(request)
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/web_app.py", line 543, in _handle
    resp = await handler(request)
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/web_middlewares.py", line 114, in impl
    return await handler(request)
  File "/src/pulpcore/pulpcore/content/authentication.py", line 48, in authenticate
    return await handler(request)
  File "/src/pulpcore/pulpcore/content/instrumentation.py", line 230, in middleware
    resp = await handler(request)
  File "/src/pulp_container/pulp_container/app/registry.py", line 158, in get_tag
    response = await downloader.run(
  File "/src/pulpcore/pulpcore/download/http.py", line 269, in run
    return await download_wrapper()
  File "/usr/local/lib/python3.8/site-packages/backoff/_async.py", line 151, in retry
    ret = await target(*args, **kwargs)
  File "/src/pulpcore/pulpcore/download/http.py", line 254, in download_wrapper
    return await self._run(extra_data=extra_data)
  File "/src/pulp_container/pulp_container/app/downloaders.py", line 79, in _run
    async with session_http_method(
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/client.py", line 1187, in __aenter__
    self._resp = await self._coro
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/client.py", line 574, in _request
    conn = await self._connector.connect(
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/connector.py", line 544, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/connector.py", line 911, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
  File "/usr/local/lib64/python3.8/site-packages/aiohttp/connector.py", line 1187, in _create_direct_connection
    raise ClientConnectorError(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host registry-1.docker.io:443 ssl:default [Name or service not known]
```

--- a/CHANGES/1499.bugfix
+++ b/CHANGES/1499.bugfix
@@ -1,0 +1,1 @@
+Made the pull-through caching machinery resilient to connection errors.

--- a/pulp_container/app/exceptions.py
+++ b/pulp_container/app/exceptions.py
@@ -1,4 +1,16 @@
-from rest_framework.exceptions import NotFound, ParseError
+from rest_framework.exceptions import APIException, NotFound, ParseError
+
+
+class BadGateway(APIException):
+    status_code = 502
+    default_detail = "Invalid response received from the upstream."
+    default_code = "BAD_GATEWAY"
+
+
+class GatewayTimeout(APIException):
+    status_code = 504
+    default_detail = "Response from the upstream timed out."
+    default_code = "GATEWAY_TIMEOUT"
 
 
 class RepositoryNotFound(NotFound):

--- a/pulp_container/app/registry.py
+++ b/pulp_container/app/registry.py
@@ -8,7 +8,7 @@ from contextlib import suppress
 from urllib.parse import urljoin
 
 from aiohttp import web
-from aiohttp.client_exceptions import ClientResponseError
+from aiohttp.client_exceptions import ClientResponseError, ClientConnectionError
 from aiohttp.web_exceptions import HTTPTooManyRequests
 from django_guid import set_guid
 from django_guid.utils import generate_guid
@@ -21,6 +21,7 @@ from pulpcore.plugin.content import Handler, PathNotResolved
 from pulpcore.plugin.models import RemoteArtifact, Content, ContentArtifact
 from pulpcore.plugin.content import ArtifactResponse
 from pulpcore.plugin.tasking import dispatch
+from pulpcore.plugin.exceptions import TimeoutException
 
 from pulp_container.app.cache import RegistryContentCache
 from pulp_container.app.models import ContainerDistribution, Tag, Blob, Manifest, BlobManifest
@@ -157,7 +158,7 @@ class Registry(Handler):
                 response = await downloader.run(
                     extra_data={"headers": V2_ACCEPT_HEADERS, "http_method": "head"}
                 )
-            except ClientResponseError:
+            except (ClientResponseError, ClientConnectionError, TimeoutException):
                 # the manifest is not available on the remote anymore
                 # but the old one is still stored in the database
                 pass

--- a/pulp_container/app/registry_api.py
+++ b/pulp_container/app/registry_api.py
@@ -12,7 +12,7 @@ import logging
 import hashlib
 import re
 
-from aiohttp.client_exceptions import ClientResponseError
+from aiohttp.client_exceptions import ClientResponseError, ClientConnectionError
 from itertools import chain
 from urllib.parse import urljoin, urlparse, urlunparse, parse_qs, urlencode
 from tempfile import NamedTemporaryFile
@@ -29,6 +29,8 @@ from pulpcore.plugin.models import Artifact, ContentArtifact, UploadChunk
 from pulpcore.plugin.files import PulpTemporaryUploadedFile
 from pulpcore.plugin.tasking import add_and_remove, dispatch
 from pulpcore.plugin.util import get_objects_for_user, get_url
+from pulpcore.plugin.exceptions import TimeoutException
+
 from rest_framework.exceptions import (
     AuthenticationFailed,
     NotAuthenticated,
@@ -55,6 +57,8 @@ from pulp_container.app.cache import (
     RegistryApiCache,
 )
 from pulp_container.app.exceptions import (
+    BadGateway,
+    GatewayTimeout,
     InvalidRequest,
     RepositoryNotFound,
     RepositoryInvalid,
@@ -1117,9 +1121,13 @@ class Manifests(RedirectsMixin, ContainerRegistryApiMixin, ViewSet):
                 # the client could request the manifest outside the docker hub pull limit;
                 # it is necessary to pass this information back to the client
                 raise Throttled()
-            else:
-                # TODO: do not mask out relevant errors, like HTTP 502
+            elif response_error.status == 404:
                 raise ManifestNotFound(reference=pk)
+            else:
+                raise BadGateway(detail=response_error.message)
+        except (ClientConnectionError, TimeoutException):
+            # The remote server is not available at the moment
+            raise GatewayTimeout()
         else:
             digest = response.headers.get("docker-content-digest")
             return models.Manifest.objects.filter(digest=digest).first()

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,3 +1,3 @@
 jsonschema>=4.4,<4.23
-pulpcore>=3.43.0,<3.55
+pulpcore>=3.46.0,<3.55
 pyjwt[crypto]>=2.4,<2.9

